protocols/autonatv2/src/request_response.rs

```diff
@@ -0,0 +1,336 @@
+            }
+            _ => Err(new_io_invalid_data_err!(
+                "invalid message type, expected dialRequest or dialDataResponse"
+            )),
+        }
+    }
+
+    pub(crate) fn into_proto(self) -> proto::Message {
+        match self {
+            Request::Dial(DialRequest { addrs, nonce }) => {
+                let addrs = addrs.iter().map(|e| e.to_vec()).collect();
+                let nonce = Some(nonce);
+                proto::Message {
+                    msg: proto::mod_Message::OneOfmsg::dialRequest(proto::DialRequest {
+                        addrs,
+                        nonce,
+                    }),
+                }
+            }
+            Request::Data(DialDataResponse { data_count }) => {
+                debug_assert!(
+                    data_count <= DATA_FIELD_LEN_UPPER_BOUND,
+                    "data_count too large"
+                );
+                static DATA: &[u8] = &[0u8; DATA_FIELD_LEN_UPPER_BOUND];
+                proto::Message {
+                    msg: proto::mod_Message::OneOfmsg::dialDataResponse(proto::DialDataResponse {
+                        data: Some(Cow::Borrowed(&DATA[..data_count])),
+                    }),
+                }
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(crate) enum Response {
+    Dial(DialResponse),
+    Data(DialDataRequest),
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct DialDataRequest {
+    pub(crate) addr_idx: usize,
+    pub(crate) num_bytes: usize,
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct DialResponse {
+    pub(crate) status: proto::mod_DialResponse::ResponseStatus,
+    pub(crate) addr_idx: usize,
+    pub(crate) dial_status: proto::DialStatus,
+}
+
+impl Response {
+    pub(crate) fn from_proto(msg: proto::Message) -> std::io::Result<Self> {
+        match msg.msg {
+            proto::mod_Message::OneOfmsg::dialResponse(proto::DialResponse {
+                status,
+                addrIdx,
+                dialStatus,
+            }) => {
+                let status = check_existence!(status)?;
+                let addr_idx = check_existence!(addrIdx)? as usize;
+                let dial_status = check_existence!(dialStatus)?;
+                Ok(Self::Dial(DialResponse {
+                    status,
+                    addr_idx,
+                    dial_status,
+                }))
+            }
+            proto::mod_Message::OneOfmsg::dialDataRequest(proto::DialDataRequest {
+                addrIdx,
+                numBytes,
+            }) => {
+                let addr_idx = check_existence!(addrIdx)? as usize;
+                let num_bytes = check_existence!(numBytes)? as usize;
+                Ok(Self::Data(DialDataRequest {
+                    addr_idx,
+                    num_bytes,
+                }))
+            }
+            _ => Err(new_io_invalid_data_err!(
+                "invalid message type, expected dialResponse or dialDataRequest"
```

Comment:
```suggestion
                "expected dialResponse or dialDataRequest"
```

Nit: The `io::ErrorKind` already says what the problem is (`InvalidData`), it is good enough if we extend it with the reason why.

Reply:
You are right